library/core/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java

```diff
@@ -1576,24 +1587,66 @@ public int outputSampleEncryptionData() {
         vectorSize = initVectorData.length;
       }
 
-      boolean subsampleEncryption = fragment.sampleHasSubsampleEncryptionTable(currentSampleIndex);
+      boolean haveSubsampleEncryptionTable =
+          fragment.sampleHasSubsampleEncryptionTable(currentSampleIndex);
+      boolean writeSubsampleEncryptionData =
+          haveSubsampleEncryptionTable | clearHeaderSize != 0;
 
       // Write the signal byte, containing the vector size and the subsample encryption flag.
-      encryptionSignalByte.data[0] = (byte) (vectorSize | (subsampleEncryption ? 0x80 : 0));
+      encryptionSignalByte.data[0] =
+          (byte) (vectorSize | (writeSubsampleEncryptionData ? 0x80 : 0));
       encryptionSignalByte.setPosition(0);
       output.sampleData(encryptionSignalByte, 1);
       // Write the vector.
       output.sampleData(initializationVectorData, vectorSize);
-      // If we don't have subsample encryption data, we're done.
-      if (!subsampleEncryption) {
+
+      if (!writeSubsampleEncryptionData) {
         return 1 + vectorSize;
       }
-      // Write the subsample encryption data.
+
+      if (!haveSubsampleEncryptionTable) {
+        // Need to synthesize subsample encryption data. The sample is fully encrypted except
+        // for the additional header that the extractor is going to prefix, so we need to write the
+        // following to output.sampleData:
+        // subsampleCount (unsigned short) = 1
+        // clearDataSizes[0] (unsigned short) = clearHeaderSize
+        // encryptedDataSizes[0] (unsigned int) = sampleSize
+        ParsableByteArray encryptionData = new ParsableByteArray(8);
+        encryptionData.data[0] = (byte)0;
+        encryptionData.data[1] = (byte)1;
+        encryptionData.data[2] = (byte)((clearHeaderSize & 0xFF00) >>> 8);
+        encryptionData.data[3] = (byte)( clearHeaderSize & 0x00FF);
+        encryptionData.data[4] = (byte)((sampleSize & 0xFF000000) >>> 24);
+        encryptionData.data[5] = (byte)((sampleSize & 0x00FF0000) >>> 16);
+        encryptionData.data[6] = (byte)((sampleSize & 0x0000FF00) >>>  8);
+        encryptionData.data[7] = (byte)( sampleSize & 0x000000FF);
+        encryptionData.setPosition(0);
+        output.sampleData(encryptionData, 8);
+        return 1 + vectorSize + 8;
```

Comment:
I'm not familiar with the usage of ParsableByteArray. The code here is a little stupid. :(
Don't know whether there is a smart one. But, this code can work. :)

Reply:
This approach looks OK to me :).